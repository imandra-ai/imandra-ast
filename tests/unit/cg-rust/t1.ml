module Fmt = CCFormat
module A = Imandra_ast

let decls_json =
  {json|

[{"loc":{"filename":"//toplevel//","line_start":2,"line_end":5},"view":["Ty",[{"params":[],"decl":["Record",[{"f":{"name":"re","id":2},"ty":{"view":["Constr",{"name":"int","id":1},[]]}},{"f":{"name":"im","id":3},"ty":{"view":["Constr",{"name":"int","id":1},[]]}}]],"name":{"name":"complex","id":0},"name_loc":{"filename":"//toplevel//","line_start":2,"line_end":2},"loc":{"filename":"//toplevel//","line_start":2,"line_end":5},"codegen_tags":[]}]]},{"loc":{"filename":"//toplevel//","line_start":7,"line_end":9},"view":["Ty",[{"params":[],"decl":["Algebraic",[{"c":{"name":"Polar","id":5},"labels":null,"args":[{"view":["Constr",{"name":"int","id":1},[]]},{"view":["Constr",{"name":"int","id":1},[]]}]},{"c":{"name":"Cartesian","id":6},"labels":null,"args":[{"view":["Constr",{"name":"complex","id":0},[]]}]}]],"name":{"name":"point","id":4},"name_loc":{"filename":"//toplevel//","line_start":7,"line_end":7},"loc":{"filename":"//toplevel//","line_start":7,"line_end":9},"codegen_tags":[]}]]},{"loc":{"filename":"//toplevel//","line_start":18,"line_end":19},"view":["Fun",{"recursive":false,"fs":[{"name":{"name":"c_plus","id":7},"pat":{"view":["P_var",{"id":{"name":"c_plus","id":7},"ty":{"view":["Arrow","",{"view":["Constr",{"name":"complex","id":0},[]]},{"view":["Arrow","",{"view":["Constr",{"name":"complex","id":0},[]]},{"view":["Constr",{"name":"complex","id":0},[]]}]}]}}],"loc":{"filename":"//toplevel//","line_start":18,"line_end":19}},"body":{"view":["Fun",{"view":["Arrow","",{"view":["Constr",{"name":"complex","id":0},[]]},{"view":["Arrow","",{"view":["Constr",{"name":"complex","id":0},[]]},{"view":["Constr",{"name":"complex","id":0},[]]}]}]},["Nolabel"],{"id":{"name":"c1","id":8},"ty":{"view":["Constr",{"name":"complex","id":0},[]]}},{"view":["Fun",{"view":["Arrow","",{"view":["Constr",{"name":"complex","id":0},[]]},{"view":["Constr",{"name":"complex","id":0},[]]}]},["Nolabel"],{"id":{"name":"c2","id":9},"ty":{"view":["Constr",{"name":"complex","id":0},[]]}},{"view":["Record",{"view":["Constr",{"name":"complex","id":0},[]]},[[{"name":"re","id":2},{"view":["Apply",{"view":["Constr",{"name":"int","id":1},[]]},{"view":["Ident",{"id":{"name":"+","id":10},"ty":{"view":["Arrow","",{"view":["Constr",{"name":"int","id":1},[]]},{"view":["Arrow","",{"view":["Constr",{"name":"int","id":1},[]]},{"view":["Constr",{"name":"int","id":1},[]]}]}]}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}},[[["Nolabel"],{"view":["Field",{"data_ty":{"view":["Constr",{"name":"complex","id":0},[]]},"ty":{"view":["Constr",{"name":"int","id":1},[]]},"f":{"name":"re","id":2},"t":{"view":["Ident",{"id":{"name":"c1","id":8},"ty":{"view":["Constr",{"name":"complex","id":0},[]]}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}],[["Nolabel"],{"view":["Field",{"data_ty":{"view":["Constr",{"name":"complex","id":0},[]]},"ty":{"view":["Constr",{"name":"int","id":1},[]]},"f":{"name":"re","id":2},"t":{"view":["Ident",{"id":{"name":"c2","id":9},"ty":{"view":["Constr",{"name":"complex","id":0},[]]}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}]]],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}],[{"name":"im","id":3},{"view":["Apply",{"view":["Constr",{"name":"int","id":1},[]]},{"view":["Ident",{"id":{"name":"+","id":10},"ty":{"view":["Arrow","",{"view":["Constr",{"name":"int","id":1},[]]},{"view":["Arrow","",{"view":["Constr",{"name":"int","id":1},[]]},{"view":["Constr",{"name":"int","id":1},[]]}]}]}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}},[[["Nolabel"],{"view":["Field",{"data_ty":{"view":["Constr",{"name":"complex","id":0},[]]},"ty":{"view":["Constr",{"name":"int","id":1},[]]},"f":{"name":"im","id":3},"t":{"view":["Ident",{"id":{"name":"c1","id":8},"ty":{"view":["Constr",{"name":"complex","id":0},[]]}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}],[["Nolabel"],{"view":["Field",{"data_ty":{"view":["Constr",{"name":"complex","id":0},[]]},"ty":{"view":["Constr",{"name":"int","id":1},[]]},"f":{"name":"im","id":3},"t":{"view":["Ident",{"id":{"name":"c2","id":9},"ty":{"view":["Constr",{"name":"complex","id":0},[]]}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}}],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}]]],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}]],null],"loc":{"filename":"//toplevel//","line_start":19,"line_end":19}}],"loc":{"filename":"//toplevel//","line_start":18,"line_end":19}}],"loc":{"filename":"//toplevel//","line_start":18,"line_end":19}},"loc":{"filename":"//toplevel//","line_start":18,"line_end":19},"name_loc":{"filename":"//toplevel//","line_start":18,"line_end":18},"codegen_tags":[]}]}]},{"loc":{"filename":"//toplevel//","line_start":11,"line_end":13},"view":["Fun",{"recursive":false,"fs":[{"name":{"name":"is_cartesian","id":11},"pat":{"view":["P_var",{"id":{"name":"is_cartesian","id":11},"ty":{"view":["Arrow","",{"view":["Constr",{"name":"point","id":4},[]]},{"view":["Constr",{"name":"bool","id":12},[]]}]}}],"loc":{"filename":"//toplevel//","line_start":11,"line_end":13}},"body":{"view":["Fun",{"view":["Arrow","",{"view":["Constr",{"name":"point","id":4},[]]},{"view":["Constr",{"name":"bool","id":12},[]]}]},["Nolabel"],{"id":{"name":"_x","id":13},"ty":{"view":["Constr",{"name":"point","id":4},[]]}},{"view":["Match",{"loc":{"filename":"//toplevel//","line_start":11,"line_end":13},"ty":{"view":["Constr",{"name":"point","id":4},[]]},"lhs":{"view":["Ident",{"id":{"name":"_x","id":13},"ty":{"view":["Constr",{"name":"point","id":4},[]]}}],"loc":{"filename":"//toplevel//","line_start":11,"line_end":13}},"bs":[{"pat":{"view":["P_construct",{"c":{"name":"Polar","id":5},"ty":{"view":["Constr",{"name":"point","id":4},[]]},"args":[{"view":["P_any",{"view":["Constr",{"name":"int","id":1},[]]}],"loc":{"filename":"//toplevel//","line_start":12,"line_end":12}},{"view":["P_any",{"view":["Constr",{"name":"int","id":1},[]]}],"loc":{"filename":"//toplevel//","line_start":12,"line_end":12}}],"lbls":null}],"loc":{"filename":"//toplevel//","line_start":12,"line_end":12}},"when_":null,"expr":{"view":["False"],"loc":{"filename":"//toplevel//","line_start":12,"line_end":12}}},{"pat":{"view":["P_construct",{"c":{"name":"Cartesian","id":6},"ty":{"view":["Constr",{"name":"point","id":4},[]]},"args":[{"view":["P_any",{"view":["Constr",{"name":"complex","id":0},[]]}],"loc":{"filename":"//toplevel//","line_start":13,"line_end":13}}],"lbls":null}],"loc":{"filename":"//toplevel//","line_start":13,"line_end":13}},"when_":null,"expr":{"view":["True"],"loc":{"filename":"//toplevel//","line_start":13,"line_end":13}}}]}],"loc":{"filename":"//toplevel//","line_start":11,"line_end":13}}],"loc":{"filename":"//toplevel//","line_start":11,"line_end":13}},"loc":{"filename":"//toplevel//","line_start":11,"line_end":13},"name_loc":{"filename":"//toplevel//","line_start":11,"line_end":11},"codegen_tags":[]}]}]}]

|json}

let decls =
  match A.Decl.list_of_of_yojson @@ Yojson.Safe.from_string decls_json with
  | Ok d -> d
  | Error e ->
    Fmt.eprintf "cannot decode decls: %s" e;
    assert false

let code = Imandra_ast_cg_rust.codegen decls
let () = Fmt.printf "rust code:@.```@.%s@.```@." code
