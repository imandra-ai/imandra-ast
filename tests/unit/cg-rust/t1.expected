rust code:
```
// generated from imandra-ast

use num_bigint::{BigInt as BigInt, Zero, One, Add};
use num_rational::BigRational as Real;

#[derive(Eq,PartialEq,Clone,Debug)]
pub struct Complex {
  pub re: BigInt,
  pub im: BigInt,
}

#[derive(Eq,PartialEq,Clone,Debug)]
pub enum Point {
  Polar(BigInt,BigInt),
  Cartesian(Complex /* complex */),
}

// TODO: fun c_plus
// TODO: fun is_cartesian

```
